File: javascript/03-copy-paste-payroll/src/payroll_calculator.js
class PayrollCalculator {
  constructor() {}

  generate_payslips(employees) {
    let payslips = [];

    for (let i = 0; i < employees.length; i++) {
      let emp = employees[i];
      let gross = 0;
      let deductions = 0;
      let net = 0;

      if (emp.type === 'fulltime') {
        gross = emp.salary / 12;
        deductions = gross * 0.25;
        if (emp.bonus) {
          gross += emp.bonus / 12;
        }
        net = gross - deductions;
      } else if (emp.type === 'parttime') {
        gross = emp.hours * emp.rate;
        deductions = gross * 0.15;
        net = gross - deductions;
      } else if (emp.type === 'contract') {
        gross = emp.flatFee;
        deductions = gross * 0.1;
        net = gross - deductions;
      }

      payslips.push({
        id: emp.id,
        name: emp.name,
        type: emp.type,
        gross: Math.round(gross * 100) / 100,
        deductions: Math.round(deductions * 100) / 100,
        net: Math.round(net * 100) / 100
      });
    }

    return payslips;
  }
}

export { PayrollCalculator };

File: golang/03-copy-paste-payroll/payroll_calculator.go
package kata

import "math"

type Employee struct {
	ID      int
	Name    string
	Type    string
	Salary  float64
	Bonus   float64
	Hours   float64
	Rate    float64
	FlatFee float64
}
type Payslip struct {
	ID         int
	Name       string
	Type       string
	Gross      float64
	Deductions float64
	Net        float64
}

type PayrollCalculator struct{}

func NewPayrollCalculator() *PayrollCalculator {
	return &PayrollCalculator{}
}

func (pc *PayrollCalculator) generate_payslips(employees []Employee) []Payslip {
	slips := []Payslip{}
	for _, emp := range employees {
		gross := 0.0
		deductions := 0.0
		net := 0.0
		if emp.Type == "fulltime" {
			gross = emp.Salary / 12
			deductions = gross * 0.25
			if emp.Bonus > 0 {
				gross += emp.Bonus / 12
			}
			net = gross - deductions
		} else if emp.Type == "parttime" {
			gross = emp.Hours * emp.Rate
			deductions = gross * 0.15
			net = gross - deductions
		} else if emp.Type == "contract" {
			gross = emp.FlatFee
			deductions = gross * 0.1
			net = gross - deductions
		}
		slips = append(slips, Payslip{emp.ID, emp.Name, emp.Type, math.Round(gross*100) / 100, math.Round(deductions*100) / 100, math.Round(net*100) / 100})
	}
	return slips
}
